src/test/scala/chiselTests/experimental/hierarchy/InstanceSpec.scala

```diff
@@ -1044,4 +1047,196 @@ class InstanceSpec extends ChiselFunSpec with Utils {
       getFirrtlAndAnnos(new HasMultipleTypeParamsInside, Seq(aspect))
     }
   }
+  describe("11: Contextual") {
+    import ContextualExamples._
+    it("11.0: Example - Set in definition, read outside definition") {
+      @instantiable
+      class AddOne extends Module {
+        @public val index = Contextual(0)
+      }
+      class Top extends Module {
+        val d0 = Definition(new AddOne)
+        d0.index should be (0)
+        val d1 = d0.withContext { case i: Int => i + 1}
+        d1.index should be (1)
+      }
+      val (chirrtl, _) = getFirrtlAndAnnos(new Top)
+    }
+    it("11.1: Example - absolute indexes of instances") {
+      import E1._
+      class AbsoluteIndexer() {
+        private var i = 0
+        def get: Int = { i += 1; i - 1 }
+      }
+      class Top extends Module {
+        val indexer = new AbsoluteIndexer()
+        val definition = Definition(new Root).withContext {
+          case NoIndex => Index(indexer.get)
+        }
+        val answers = Map(
+          ("~Top|Root/i0:Middle/i0:Leaf" -> Index(0)),
+          ("~Top|Root/i0:Middle/i1:Leaf" -> Index(1)),
+          ("~Top|Root/i1:Middle/i0:Leaf" -> Index(2)),
+          ("~Top|Root/i1:Middle/i1:Leaf" -> Index(3))
+        )
+        Select.allInstancesOf[Leaf](definition).foreach { case l: Instance[Leaf] =>
+          l.index should be (answers(l.toTarget.toString))
+        }
```

Comment:
As we discussed IRL, I am concerned about the fact that this doesn't really give a stable result if you run it twice. I think we talked about storing the stable result in an auxiliary data structure, but then how is this better/different than building an Aspect over this thing and putting the mapping from Instance -> Index into the auxiliary data structure?